File: src/test/java/de/spries/fleetcommander/model/ai/DefaultBuildingStrategyTest.java
package de.spries.fleetcommander.model.ai;

import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

import java.util.Arrays;
import java.util.List;

import org.junit.Before;
import org.junit.Test;
import org.mockito.InOrder;
import org.mockito.Mockito;

import de.spries.fleetcommander.model.facade.PlayerSpecificPlanet;
import de.spries.fleetcommander.model.facade.PlayerSpecificUniverse;

public class DefaultBuildingStrategyTest {

	private PlayerSpecificUniverse universe;
	private BuildingStrategy buildingStrategy;
	private PlayerSpecificPlanet homePlanet;
	private PlayerSpecificPlanet closePlanet;
	private PlayerSpecificPlanet distantPlanet;

	@Before
	public void setUp() throws Exception {
		buildingStrategy = new DefaultBuildingStrategy();
		universe = mock(PlayerSpecificUniverse.class);
		homePlanet = mock(PlayerSpecificPlanet.class);
		closePlanet = mock(PlayerSpecificPlanet.class);
		distantPlanet = mock(PlayerSpecificPlanet.class);

		doReturn(homePlanet).when(universe).getHomePlanet();
		List<PlayerSpecificPlanet> planets = Arrays.asList(closePlanet, distantPlanet, homePlanet);
		doReturn(planets).when(universe).getPlanets();

		doReturn(0.).when(homePlanet).distanceTo(homePlanet);
		doReturn(1.).when(closePlanet).distanceTo(homePlanet);
		doReturn(2.).when(distantPlanet).distanceTo(homePlanet);

		doReturn(true).when(homePlanet).isInhabitedByMe();
		doReturn(true).when(closePlanet).isInhabitedByMe();
		doReturn(true).when(distantPlanet).isInhabitedByMe();
	}

	@Test
	public void buildsNoFactoriesWhenPlayerCannotBuild() throws Exception {
		doReturn(false).when(homePlanet).canBuildFactory();
		buildingStrategy.buildFactories(universe);
		verify(homePlanet, never()).buildFactory();
	}

	@Test
	public void buildsFactoriesWhenPlayerCanBuild() throws Exception {
		doReturn(true).doReturn(true).doReturn(false).when(homePlanet).canBuildFactory();
		buildingStrategy.buildFactories(universe);
		verify(homePlanet, times(2)).buildFactory();
	}

	@Test
	public void buildsFactoriesOnPlanetsCloseToHomePlanetFirst() {
		doReturn(true).doReturn(false).when(homePlanet).canBuildFactory();
		doReturn(true).doReturn(false).when(closePlanet).canBuildFactory();
		doReturn(true).doReturn(false).when(distantPlanet).canBuildFactory();

		buildingStrategy.buildFactories(universe);

		InOrder inOrder = Mockito.inOrder(homePlanet, closePlanet, distantPlanet);
		inOrder.verify(homePlanet).buildFactory();
		inOrder.verify(closePlanet).buildFactory();
		inOrder.verify(distantPlanet).buildFactory();
	}

}

File: src/test/kotlin/de/spries/fleetcommander/model/ai/behavior/DefaultBuildingStrategyTest.kt
package de.spries.fleetcommander.model.ai.behavior

import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import de.spries.fleetcommander.model.facade.PlayerSpecificPlanet
import de.spries.fleetcommander.model.facade.PlayerSpecificUniverse
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito
import org.mockito.Mockito.doReturn
import org.mockito.Mockito.never
import org.mockito.Mockito.times
import org.mockito.Mockito.verify

class DefaultBuildingStrategyTest {

    private lateinit var universe: PlayerSpecificUniverse
    private lateinit var buildingStrategy: BuildingStrategy
    private lateinit var homePlanet: PlayerSpecificPlanet
    private lateinit var closePlanet: PlayerSpecificPlanet
    private lateinit var distantPlanet: PlayerSpecificPlanet

    @Before
    @Throws(Exception::class)
    fun setUp() {
        buildingStrategy = DefaultBuildingStrategy()
        universe = mock()
        homePlanet = mock()
        closePlanet = mock()
        distantPlanet = mock()

        doReturn(homePlanet).`when`(universe).getHomePlanet()
        val planets = listOf(closePlanet, distantPlanet, homePlanet)
        doReturn(planets).`when`(universe).getPlanets()

        whenever(homePlanet.distanceTo(homePlanet)).thenReturn(0.0)
        whenever(closePlanet.distanceTo(homePlanet)).thenReturn(1.0)
        whenever(distantPlanet.distanceTo(homePlanet)).thenReturn(2.0)

        doReturn(true).`when`(homePlanet).isInhabitedByMe()
        doReturn(true).`when`(closePlanet).isInhabitedByMe()
        doReturn(true).`when`(distantPlanet).isInhabitedByMe()
    }

    @Test
    @Throws(Exception::class)
    fun buildsNoFactoriesWhenPlayerCannotBuild() {
        whenever(homePlanet.canBuildFactory()).thenReturn(false)
        buildingStrategy.buildFactories(universe)
        verify(homePlanet, never()).buildFactory()
    }

    @Test
    @Throws(Exception::class)
    fun buildsFactoriesWhenPlayerCanBuild() {
        whenever(homePlanet.canBuildFactory()).thenReturn(true).thenReturn(true).thenReturn(false)
        buildingStrategy.buildFactories(universe)
        verify(homePlanet, times(2)).buildFactory()
    }

    @Test
    fun buildsFactoriesOnPlanetsCloseToHomePlanetFirst() {
        whenever(homePlanet.canBuildFactory()).thenReturn(true).thenReturn(false)
        whenever(closePlanet.canBuildFactory()).thenReturn(true).thenReturn(false)
        whenever(distantPlanet.canBuildFactory()).thenReturn(true).thenReturn(false)

        buildingStrategy.buildFactories(universe)

        val inOrder = Mockito.inOrder(homePlanet, closePlanet, distantPlanet)
        inOrder.verify(homePlanet).buildFactory()
        inOrder.verify(closePlanet).buildFactory()
        inOrder.verify(distantPlanet).buildFactory()
    }

}
